File: frontend/src/utils/projectAgentDefaults.js
// frontend/src/utils/projectAgentDefaults.js
//
// Per-project agent-settings defaults: frontend chain resolver. Given the
// projects already in the store, walk a project's ancestor chain (worktree
// main repo first, else nearest path ancestor, recursively) and resolve —
// field by field — the inherited agent-settings defaults and the inherited
// default provider. Resolution is a CREATION-TIME concern only: this resolver
// pre-fills a new draft with concrete values (frozen at launch — the snapshot
// model, see docs/plans/2026-06-09-project-agent-defaults-design.md §4), and
// also drives the display baseline (popover diff-marking and reset targets).
// The backend never re-resolves the chain for a running session; its mirror
// of this resolver (src/twicc/project_agent_defaults.py) runs only when the
// CLI creates a session (same materialize-at-creation semantics). Keep the
// two resolvers in sync.
//
// Unlike trust (utils/trust.js) there is NO propagation gate: every ancestor is
// visited and, for each field, the first non-null value wins. The chain is also
// UNFILTERED — an ancestor with no own defaults is still a step up, because a
// different field may be set higher.
//
// Each project carries: { id, directory, worktree_of, default_provider,
// default_agent_settings }. `default_agent_settings` is
// { <provider>: { <AgentSettings field>: value|null } } or null; field names
// are the canonical wire names (selected_model, thinking_enabled, effort,
// permission_mode, context_max, claude_in_chrome, fast_mode).

function segments(path) {
    return path.split('/').filter(Boolean)
}

// True iff `ancestor` is a strict directory prefix of `descendant`, compared
// segment-by-segment (so /a/b is NOT an ancestor of /a/bc). Both are realpath'd.
function isPathAncestor(ancestor, descendant) {
    const a = segments(ancestor)
    const d = segments(descendant)
    if (a.length >= d.length) return false
    for (let i = 0; i < a.length; i++) {
        if (a[i] !== d[i]) return false
    }
    return true
}

// Nearest (longest-path) ancestor — UNFILTERED: any registered project that is a
// strict path-prefix qualifies, regardless of whether it has its own defaults.
function nearestPathAncestor(project, projectsById) {
    let best = null
    let bestLen = -1
    for (const id in projectsById) {
        const q = projectsById[id]
        if (q.id === project.id || !q.directory) continue
        if (isPathAncestor(q.directory, project.directory)) {
            const depth = segments(q.directory).length
            if (depth > bestLen) {
                best = q
                bestLen = depth
            }
        }
    }
    return best
}

/**
 * Ordered ancestor chain [self, parent, grandparent, ...]. At each node the
 * parent is its `worktree_of` main repo if set, else its nearest path ancestor.
 * Cycle-guarded against pathological `worktree_of` loops.
 */
export function ancestorChain(projectId, projectsById) {
    const chain = []
    const seen = new Set()
    let node = projectsById[projectId]
    while (node && !seen.has(node.id)) {
        chain.push(node)
        seen.add(node.id)
        if (node.worktree_of) {
            node = projectsById[node.worktree_of]
        } else if (node.directory) {
            node = nearestPathAncestor(node, projectsById)
        } else {
            node = null
        }
    }
    return chain
}

/**
 * Resolve the inherited per-project agent-settings defaults for a provider.
 * ``options.includeSelf`` defaults to true. Set it to false when resolving the
 * inheritance baseline for a project whose own sparse values are being edited.
 *
 * @returns {Object} a { field: value } map holding only the fields the chain
 *   actually sets; a missing field means "inherit further / fall back to the
 *   global default" (the caller applies that global fallback).
 */
export function resolveProjectAgentDefaults(projectId, provider, projectsById, { includeSelf = true } = {}) {
    const resolved = {}
    if (!projectId || !provider) return resolved
    const chain = ancestorChain(projectId, projectsById)
    for (const node of includeSelf ? chain : chain.slice(1)) {
        const bundle = (node.default_agent_settings && node.default_agent_settings[provider]) || {}
        for (const field in bundle) {
            const value = bundle[field]
            // First non-null wins (nearest ancestor); an explicit null means
            // "inherit", so it does not block a value set further up.
            if (value != null && !(field in resolved)) {
                resolved[field] = value
            }
        }
    }
    return resolved
}

/**
 * Resolve the inherited default provider for a project (worktree/path chain).
 * @returns {(string|null)} a provider value, or null when nothing in the chain
 *   sets one (the caller falls back to the global default provider).
 */
export function resolveProjectDefaultProvider(projectId, projectsById) {
    if (!projectId) return null
    for (const node of ancestorChain(projectId, projectsById)) {
        if (node.default_provider) return node.default_provider
    }
    return null
}

export function resolveDraftProvider(projectId, projectsById, globalDefault, explicitProvider = null) {
    return explicitProvider
        ?? resolveProjectDefaultProvider(projectId, projectsById)
        ?? globalDefault
}

File: src/twicc/project_agent_defaults.py
"""Per-project agent-settings defaults: backend chain resolution (creation-only).

Backend mirror of ``frontend/src/utils/projectAgentDefaults.js``. Given the
defaults-relevant slice of every project, walk a project's ancestor chain
(worktree main repo first, else nearest registered path ancestor, recursively)
and resolve — field by field — the inherited agent-settings defaults and the
inherited default provider.

This resolver runs at **session creation only** (the CLI ``create-session``
materializes the resolved values into the payload, exactly like the frontend
freezes them onto a draft — see
``docs/plans/2026-06-09-project-agent-defaults-design.md`` §4, "snapshot at
creation"). It must NEVER be wired into the per-turn resolution sites
(``resolve_agent_settings`` and its callers): re-resolving a running session
against the project chain is precisely the option-B bug the design removed.

Unlike trust (``twicc/trust.py``) there is NO propagation gate: every ancestor
is visited and, for each field, the first non-null value wins. The chain is
also UNFILTERED — an ancestor with no own defaults is still a step up, because
a different field may be set higher.
"""

from __future__ import annotations

import os
from typing import NamedTuple


class ProjectDefaultsRow(NamedTuple):
    """The defaults-relevant slice of a ``Project``, loaded once for resolution."""

    id: str
    directory: str | None
    worktree_of_id: str | None
    default_provider: str | None
    # ``{ "<provider>": { "<AgentSettings field>": value } }`` or ``None``.
    default_agent_settings: dict | None


def _segments(path: str) -> list[str]:
    """Normalized, non-empty path segments (pure — same rules as twicc/trust.py)."""
    return [s for s in os.path.normpath(path).split(os.sep) if s]


def _is_path_ancestor(ancestor: str, descendant: str) -> bool:
    """True iff *ancestor* is a strict directory prefix of *descendant*.

    Compared segment-by-segment (not ``str.startswith``) so ``/a/b`` is not an
    ancestor of ``/a/bc``. Both paths are assumed already realpath-normalized
    (``Project.directory`` is stored realpath'd); ``normpath`` only tidies them.
    """
    a, d = _segments(ancestor), _segments(descendant)
    return len(a) < len(d) and d[: len(a)] == a


def _nearest_path_ancestor(
    target: ProjectDefaultsRow, rows: list[ProjectDefaultsRow]
) -> ProjectDefaultsRow | None:
    """The nearest (longest-path) registered ancestor of *target* — UNFILTERED.

    Any registered project that is a strict path-prefix qualifies, regardless
    of whether it carries its own defaults (a different field may be set on it,
    or further up through it).
    """
    if not target.directory:
        return None
    best: ProjectDefaultsRow | None = None
    best_len = -1
    for row in rows:
        if row.id == target.id or not row.directory:
            continue
        if _is_path_ancestor(row.directory, target.directory):
            depth = len(_segments(row.directory))
            if depth > best_len:
                best, best_len = row, depth
    return best


def ancestor_chain(
    target: ProjectDefaultsRow, rows: list[ProjectDefaultsRow]
) -> list[ProjectDefaultsRow]:
    """Ordered ancestor chain ``[self, parent, grandparent, ...]``.

    At each node the parent is its ``worktree_of`` main repo if set, else its
    nearest registered path ancestor. Cycle-guarded against pathological
    ``worktree_of`` loops.
    """
    by_id = {r.id: r for r in rows}
    chain: list[ProjectDefaultsRow] = []
    seen: set[str] = set()
    node: ProjectDefaultsRow | None = target
    while node is not None and node.id not in seen:
        chain.append(node)
        seen.add(node.id)
        if node.worktree_of_id:
            node = by_id.get(node.worktree_of_id)
        elif node.directory:
            node = _nearest_path_ancestor(node, rows)
        else:
            node = None
    return chain


def resolve_agent_defaults(
    target: ProjectDefaultsRow, provider: str, rows: list[ProjectDefaultsRow]
) -> dict:
    """Resolve the inherited per-project agent-settings defaults for *provider*.

    Returns a ``{field: value}`` map holding only the fields the chain actually
    sets; a missing field means "fall back to the global default" (the caller
    applies that global fallback). Field keys are the canonical wire names,
    including the default-shaping ``permission_mode_if_untrusted``.
    """
    resolved: dict = {}
    for node in ancestor_chain(target, rows):
        bundle = (node.default_agent_settings or {}).get(provider) or {}
        for field, value in bundle.items():
            # First non-null wins (nearest ancestor); an explicit null means
            # "inherit", so it does not block a value set further up.
            if value is not None and field not in resolved:
                resolved[field] = value
    return resolved


def resolve_default_provider(
    target: ProjectDefaultsRow, rows: list[ProjectDefaultsRow]
) -> str | None:
    """Resolve the inherited default provider (worktree/path chain).

    Returns a provider value, or ``None`` when nothing in the chain sets one
    (the caller falls back to the global default provider).
    """
    for node in ancestor_chain(target, rows):
        if node.default_provider:
            return node.default_provider
    return None


# --- DB-facing helpers ----------------------------------------------------


def load_defaults_rows() -> list[ProjectDefaultsRow]:
    """Load the defaults-relevant slice of every project (sync DB access)."""
    from twicc.core.models import Project

    return [
        ProjectDefaultsRow(
            id=row["id"],
            directory=row["directory"],
            worktree_of_id=row["worktree_of_id"],
            default_provider=row["default_provider"],
            default_agent_settings=row["default_agent_settings"],
        )
        for row in Project.objects.values(
            "id", "directory", "worktree_of_id",
            "default_provider", "default_agent_settings",
        )
    ]


def _target_row(
    project_id: str, directory: str | None, rows: list[ProjectDefaultsRow]
) -> ProjectDefaultsRow:
    """The target's row, synthesized from the directory hint when unregistered.

    A CLI ``create-session`` may point at a directory with no ``Project`` row
    yet (the server registers it at creation). A synthetic row (no own
    defaults, no worktree link) still lets path-ancestor inheritance apply —
    e.g. a brand-new sub-directory of a project that sets defaults.
    """
    target = next((r for r in rows if r.id == project_id), None)
    if target is not None:
        return target
    return ProjectDefaultsRow(
        id=project_id,
        directory=directory,
        worktree_of_id=None,
        default_provider=None,
        default_agent_settings=None,
    )


def resolve_project_agent_defaults(
    project_id: str, provider: str, *, directory: str | None = None
) -> dict:
    """Resolve one project's inherited agent defaults against the live DB (sync)."""
    rows = load_defaults_rows()
    return resolve_agent_defaults(_target_row(project_id, directory, rows), provider, rows)


def resolve_project_default_provider(
    project_id: str, *, directory: str | None = None
) -> str | None:
    """Resolve one project's inherited default provider against the live DB (sync)."""
    rows = load_defaults_rows()
    return resolve_default_provider(_target_row(project_id, directory, rows), rows)
